src/main/java/bio/terra/service/filedata/google/gcs/GcsPdao.java

```diff
@@ -73,6 +76,10 @@
       List.of(
           "openid", "email", "profile", "https://www.googleapis.com/auth/devstorage.full_control");
 
+  // Cache of pet service account tokens keys on a given user's actual access_token
+  private final Map<String, String> petAccountTokens =
+      Collections.synchronizedMap(new PassiveExpiringMap<>(30, TimeUnit.MINUTES));
```

Comment:
Do we know the expiration time of each token? I worry that a token might expire in like 29 minutes before being added to the cache and then only being invalidated after 30 mins. Something like https://bitbucket.org/ccperkins/fecache/src/master/ could be used to expire individual tokens based on their own expiration time. There's probably a native Java way to do this, but just a thought I had.

Reply:
So what we get back from Sam doesn't encode the expiration.  We'd have to make another call to google to get the expiration which I'd love to avoid.  The token is generated the same way as the user's main auth token (which by definition already exists) so it should expire before trying to use the cache (which would then require a new auth token to be obtained by the user, which would cause a new entry in the cache).

Reply:
Oh cool. Definitely worth avoiding another call for each token.

Reply:
Yeah, I'm been poking a little so see if we can use a google lib to do the lookup.  If it's infrequent, it may not be terrible

Reply:
Confirming that this would require pulling in more libraries.  I'll poke a bit more for a potential followon but I propose keeping this approach for now to unblock folks